react-backend/client/src/cb-admin/pages/VisitorDetails.js

```diff
@@ -0,0 +1,225 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import styled from 'styled-components';
+import moment from 'moment';
+import { assocPath, filter, invertObj, map, project, sortBy } from 'ramda';
+import { FlexContainerCol, FlexContainerRow } from '../../shared/components/layout/base';
+import { Heading, Link } from '../../shared/components/text/base';
+import { Form as Fm } from '../../shared/components/form/base';
+import LabelledSelect from '../../shared/components/form/LabelledSelect';
+import TranslucentTable from '../components/TranslucentTable';
+import { Visitors } from '../../api';
+
+
+const Nav = styled.nav`
+  display: flex;
+  align-items: center;
+  justify-content: space-between;
+  flex: 1;
+`;
+
+const FlexItem = styled.div`
+  flex: ${props => props.flex || '1'};
+  height: 100%;
+`;
+
+const Form = Fm.extend`
+  width: 100%;
+`;
+
+const Row = FlexContainerRow.extend`
+  align-content: center;
+  align-items: flex-start;
+  flex: 3;
+`;
+
+const HyperLink = Link.extend`
+  flex: ${props => props.flex || '1'};
+`;
+
+const FormSection = styled.section`
+  width: 20%;
+`;
+
+
+const keyMap = {
+  visit_id: null,
+  visitor_id: 'Visitor ID',
+  gender: 'Gender',
+  yob: 'Year of birth',
+  activity: 'Activities',
+  visit_date: 'Date of visit',
+};
+
+const colToState = invertObj(keyMap);
+
+const columns = Object.values(keyMap).filter(Boolean);
+
+const sortOptions = [{ key: '0', value: '' }].concat(columns.map((col, i) => ({ key: `${i + 1}`, value: col })));
+
+const genderOptions = [
+  { key: '0', value: '' },
+  { key: '1', value: 'Male' },
+  { key: '2', value: 'Female' },
+  { key: '3', value: 'Prefer not to say' },
+];
+
+const ageOptions = [
+  { key: '0', value: '' },
+  { key: '1', value: '0-17' },
+  { key: '2', value: '18-35' },
+  { key: '3', value: '35-50' },
+  { key: '4', value: '51-69' },
+  { key: '5', value: '70+' },
+];
+
+
+const sortItems = (items, order, field) =>
+  sortBy(id => items[id][field], order);
+
+const currentYear = () => (new Date()).getFullYear();
+
+export default class VisitorDetailsPage extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      visits: {
+        items: {},
+        order: [],
+      },
+      sort: '',
+      genderFilter: '',
+      ageFilter: '',
+      errors: {},
+    };
+  }
+
+  componentDidMount() {
+    const cbAdminToken = this.props.auth;
+
+    Visitors.get(cbAdminToken, { withVisits: true })
+      .then((res) => {
+        this.props.updateAdminToken(res.headers.authorization);
+        const visits = res.data.result;
+
+        const order = visits.map(v => v.visit_id);
+        const items = visits.reduce((acc, v) => {
+          acc[v.visit_id] = { ...v, display: true };
+          return acc;
+        }, {});
+
+        this.setState(assocPath(['visits'], { order, items }));
```

Comment:
What's the benefit of using `assocPath`?

Reply:
Using the functional form of `setState` is recommended because state updates are asynchronous.

Also (the real reason): I temporarily forgot how to do this with object spreading:
```js
this.setState(state => ({ ...state, visits: { ...state.visits, order, items } }));
```
Interested to know which you think is more readable (assuming the reader understands what `assocPath` does).